doc/dev/mds_internals/quiesce.rst

```diff
@@ -0,0 +1,113 @@
+communication exists between clients. This communication can lead to clients
+wrongly believing they've reached a checkpoint that is mutually recoverable to
+via a snapshot.
+
+
+Mechanism
+---------
+
+The MDS quiesces I/O using a new ``quiesce_path`` internal request that obtains
+appropriate locks on the root of a tree and then launches a series of
+sub-requests for locking other inodes in the tree. The locks obtained will
+force clients to release caps and in-progress client/MDS requests to complete.
+
+The sub-requests launched are ``quiesce_inode`` internal requests. These will
+obtain "cap-related" locks which control capability state, including the
+``filelock``, ``authlock``, ``linklock``, and ``xattrlock``. Additionally, the
+new local lock ``quiescelock`` is acquired. More information on that lock in
+the next section.
+
+Locks that are not cap-related are skipped because they do not control typical
+and durable metadata state. Additionally, only Capabilities can give a client
+local control of a file's metadata or data.
+
+Once all locks have been acquired, the cap-related locks are released and the
+``quiescelock`` is relied on to prevent issuing Capabilities to clients for the
+cap-related locks. This is controlled primarily by ``CInode:get_caps_*``
+methods. Releasing these locks is necessary to allow other ranks with the
+replicated inode to quiesce without lock state transitions resulting in
+deadlock. For example, a client wanting ``Xx`` on an inode will trigger a
+``xattrlock`` in ``LOCK_SYNC`` state to transition to ``LOCK_SYNC_EXCL``.  That
+state would not allow another rank to acquire ``xattrlock`` for reading,
+thereby creating deadlock, subject to quiesce expiration. (Quiesce cannot
+complete until all ranks quiesce the tree.)
+
+Finally, if the inode is a directory, the ``quiesce_inode`` operation traverses
+all directory fragments and issues new ``quiesce_inode`` requests for any child
+inodes.
+
+
+Inode Quiescelock
+-----------------
+
+The ``quiescelock`` is a new local lock for inodes which supports quiescing
+I/O.  It is a type of superlock where every client or MDS operation which
+requires a wrlock or xlock on a "cap-related" inode lock will also implicitly
+acquire a wrlock on the ``quiescelock``.
+
+.. note:: A local lock supports multiple writers and only one exclusive locker. No read locks.
+
+During normal operation in the MDS, the ``quiescelock`` is never held except
+for writing. However, when a subtree is quiesced, the ``quiesce_inode``
+internal operation will hold ``quiescelock`` exclusively for the entire
+lifetime of the ``quiesce_inode`` operation. This will deny the **new**
+acquisition of any other cap-related inode lock.  The ``quiescelock`` must be ordered
+before all other locks (see ``src/include/ceph_fs.h`` for ordering) in order to
+act as this superlock.
+
+One primary reason for this ``quiescelock`` is to prevent a client request from
```

Comment:
I think this paragraph is outdated

Reply:
I don't think so? The `quiescelock` still causes an mdr to drop all locks if acquisition fails. We do not want a client to block waiting for the quiescelock while holding a rdlock on a parent's snaplock.

Reply:
I updated some of the wording in the following paragraph however.